nixos/pam: assemble rules from modular configuration

Majiir opened the pull request:
## Description

Refactors the `security.pam` module to assemble `/etc/pam.d` files from modular rule configuration. Rules can now be modified, added, toggled, and reordered.

## Problem statement

The existing module is inflexible and does not allow a user to reconfigure the PAM stack:

- Module arguments cannot be changed if there is not already a corresponding option.
- Rule `control` settings cannot be changed.
- Rules cannot be reordered.
- New rules can only be added at the beginning or end of the stack.

See [this Discourse thread](https://discourse.nixos.org/t/what-do-you-want-from-pam-security-pam-in-nixos/33265) for more discussion.

## Design

A rule set is defined for each `service` and `type` (`account`, `auth`, `password` or `session`). A rule set is now expressed as an attribute set of named rules:

```nix
{
  security.pam.services.su.rules.auth = {
    wheel = {
      enable = true;
      control = "required";
      modulePath = "pam_wheel.so";
      order = config.security.pam.services.su.rules.auth.unix.order - 10;
    };
  };
}
```

The rule name (attribute key) can be used to merge multiple configurations of the same rule (e.g. to override a built-in rule).

Rules have an `order` option which determines the order the rules are written to the PAM config files. Built-in rules are automatically assigned an `order` value based on their hard-coded position. Some kind of ordering mechanism is necessary because the rules are now configured in an attribute set.

The new config options are hidden and marked experimental so that maintainers may freely modify the option design or the ordering of rules.

## Example

```nix
{
  security.pam.services = let
    serviceCfg = service: {
      rules.auth = {
        unix = {
          control = lib.mkForce "requisite";
        };
        rssh = {
          order = config.security.pam.services.${service}.rules.auth.unix.order + 10;
          control = "sufficient";
          modulePath = "${pkgs.pam_rssh}/lib/libpam_rssh.so";
          settings.auth_key_file = "/etc/rssh/authorized_keys";
        };
      };
    };
  in lib.flip lib.genAttrs serviceCfg [
    "sudo"
    "su"
  ];
}
```

This example configures the PAM stack in ways not currently possible:

- The `pam_rssh` module is enabled and integrated into the existing service rules.
- The `pam_unix` rule is changed from `requires` to `requisite`.
- The `pam_rssh` rule is ordered after the `pam_unix` rule.
- The `pam_rssh` module is passed an additional `auth_key_file` argument.

Authentication section of the resulting `/etc/pam.d/sudo` file:

```
# Authentication management.
auth requisite pam_unix.so likeauth try_first_pass # 11600
auth sufficient /nix/store/cn5lccdcx38k81wjz42v63k8c82y2kkq-pam_rssh-1.1.0/lib/libpam_rssh.so auth_key_file=/etc/rssh/authorized_keys # 11610
auth required pam_deny.so # 12400
```

## Review notes

- Changes are best reviewed commit-by-commit.
- Many of the commits can be reproduced with regex substitution.
- None of the commits change the behavior of the resulting PAM rules.

## Related work

- https://github.com/NixOS/nixpkgs/issues/90640
- https://github.com/NixOS/nixpkgs/pull/105098
- https://github.com/NixOS/nixpkgs/pull/105319

Since previous attempts lost steam and were abandoned, I'm hoping we can merge this smaller rework and iterate on the module design in subsequent PRs.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] armv7l-linux (cross)
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - [x] `pam-file-contents`
    - [x] `pam-oath-login`
    - [x] `pam-u2f`
    - [x] `pam-ussh`
    - [x] `pam-zfs-key`
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Majiir committed:
nixos/pam: automatically populate rule type

Eliminates a redundancy between the 'rules' suboptions and the type
specified in each rule.

We eventually want to give each rule a name so that we can merge config
overrides. The PAM name is a natural choice for rule name, but a PAM is
often used in multiple rule types. Organizing rules by type and rule
name avoids name collisions.

Majiir committed:
nixos/pam: extract header comments

Unblocks converting the rules from one big string to a rich data
structure.

Majiir committed:
nixos/pam: define rules as submodules

Allows us to decompose rules into multiple fields that we later format
as textual rules. Eventually allows users to override individual fields.

Majiir committed:
nixos/pam: give each rule a name

These names are internal identifiers. They will be used as keys so that
users can reconfigure rules by merging a rule config with the same name.
The name is arbitrary. The built-in rules are named after the PAM where
practical.

Majiir committed:
nixos/pam: extract control field

Majiir committed:
nixos/pam: extract modulePath field

Majiir committed:
nixos/pam: remove empty text fields

Majiir committed:
nixos/pam: convert rules to attrs, add order field

Makes it possible to override properties of a rule by name. Introduces
an 'order' field that can be overridden to change the sequence of rules.

For now, the order value for each built-in rule is derived from its
place in the hardcoded list of rules.

Majiir committed:
nixos/pam: extract args field

Module arguments have common escaping rules for all PAMs.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-do-you-want-from-pam-security-pam-in-nixos/33265/1

Majiir committed:
nixos/pam: generate apparmor includes from rules

Removes redundant config from the module. Fixes a bug where some modules
(e.g. ussh) were added to apparmor even though they had no rules enabled.

Majiir committed:
nixos/pam: clean up rules

Makes the rules more uniform in structure and style. This makes it
easier to automate subsequent commits. No behavior changes.

Majiir committed:
nixos/pam: split rule lists into individual rules

Majiir committed:
nixos/pam: add settings option for common argument styles

Adds easily overrideable settings for the most common PAM argument
styles. These are:

- Flag (e.g. "use_first_pass"): rendered for true boolean values. false
  values are ignored.

- Key-value (e.g. "action=validate"): rendered for non-null, non-boolean
  values.

Most PAM arguments can be configured this way. Others can still be
configured with the 'args' option.

Majiir committed:
nixos/pam: add maintainer

Majiir commented:
- Improved commit readability.
- Renamed a couple of the new options.
- Fixed trailing whitespace issues.

**This is ready for review!** I've been running this on my daily driver without issues. The commits are easy to get through, I promise. Take a look and leave your comments! :bowing_man:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-do-you-want-from-pam-security-pam-in-nixos/33265/8

liff commented:
A quick thought (I haven’t spent time considering all the implications), but instead of a numerical ordering, would it make sense to use a DAG like [home-manager does](https://github.com/nix-community/home-manager/blob/a88df2fb101778bfd98a17556b3a2618c6c66091/docs/writing-modules.adoc#option-types)?

With something home-manager’s machinery the `su` example would become:

```nix
{
  security.pam.services.su.types.auth.rules = {
    unix = {
      control = lib.mkForce "requisite";
    };
    rssh = dag.entryAfter [ "unix" ] {
      control = "sufficient";
      modulePath = "${pkgs.pam_rssh}/lib/libpam_rssh.so";
      settings.auth_key_file = "/etc/rssh/authorized_keys";
    };
  };
}
```

This might make using entries with `[value=N]` controls somewhat unreliable, but I suspect that’s the case even with numbered ordering.

rissson commented:
> A quick thought (I haven’t spent time considering all the implications), but instead of a numerical ordering, would it make sense to use a DAG like [home-manager does](https://github.com/nix-community/home-manager/blob/a88df2fb101778bfd98a17556b3a2618c6c66091/docs/writing-modules.adoc#option-types)?
> 
> This might make using entries with `[value=N]` controls somewhat unreliable, but I suspect that’s the case even with numbered ordering.

During my original rework I decided against doing something like this especially because it may end with an inconsistent ordering of options. I quite like the way this PR deals with it by recommending referencing existing entries' order.

lorenzleutgeb commented:
I also think that a DAG would be better.
 1. The fact that one *can* specify a constant for `order` is a footgun.
 1. `order` defaulting to `1000 + i * 10` raises eyebrows:
  a. What if I refer to more than 10 rules to be in order after each other (using the `... + 1` mechanism suggested)? Looks like I would get two rules with the same `order` value.
  b. What if I create more than 1000 / 10 rules? Again, it looks like I would get two rules with the same `order` value.
 1. How are rules with the same `order` value placed? If there's no tie breaking, I fear that this will be done non-deterministically, or at least the user loses control of whichever deterministic process does the ordering.

@rissson could you please explain or give an example for an "inconsistent ordering"? Generally, if the order implied by the DAG is not total, then yes, there are multiple ways of manifesting the rules in a sequence. But in that case, one can add an edge. And some deterministic tie-breaking should be considered. However, you get similar (the same?) issues with just numbering (see (2.) and (3.) above).

Majiir commented:
Thanks for linking the home-manager reference. I think the DAG option is interesting!

Both DAG and `order` face ambiguous ordering scenarios, and I think both can be made to tie-break on the rule name. In the case of `order`, I think this happens already because `sort` is a stable sort.

With `order`, we can define well-known values, for example `1000` could mean "first pass", 2000 could mean "2FA", etc. With DAG, we can do something similar by ordering rules before/after some well-defined labels (almost like systemd targets).

I don't know how the DAG option handles config overrides. Is it practical for users to reorder rules? If there is a graph `a -> b -> c`, then setting `c -> b` would create a cycle, so the user would have to somehow unset `b -> c` in order to get `a -> c -> b`. Reordering like this is simpler with `order` at first glance.

I'm not worried about exhausting the rule space with `order`. We can make the spacing 100 or 1000 and make it arbitrarily unlikely that someone will declare that many PAM rules.

> What if I create more than 1000 / 10 rules? Again, it looks like I would get two rules with the same order value.

The values don't wrap around. If there are many rules, the value will just get arbitrarily large. The `order` for a rule defined in the hardcoded list in `pam.nix` is `1000 + i * 10`, but the [default for the submodule](https://github.com/NixOS/nixpkgs/blob/57b5f1e3521e8626573eed9624b432818d345741/nixos/modules/security/pam.nix#L490) is simply `1000`.

> This might make using entries with `[value=N]` controls somewhat unreliable, but I suspect that’s the case even with numbered ordering.

Yes, I think it's a (mild) problem with both. It can be solved with `substack` rules so that `[value=N]` controls are never needed.

---

My thinking is that _for now_, it doesn't matter much which one we choose. The new options are all hidden from the manual, so we can freely change the option interface in a follow-up PR. It should be easy to find and update usages inside nixpkgs.

I lean toward sticking with `order` because (a) it's more clear how rules may be reordered, and (b) it looks like the DAG option involves pulling some code from home-manager into nixpkgs lib. I would also like to experiment with the user experience of the DAG option to make sure it supports the main use cases.

Overall, I'm aiming to make a series of PRs to overhaul `pam.nix`, so it's not important that this one gets everything perfect.

**Questions for all:**

1. Do you agree that it's okay to stick with `order` for at least this first round of changes?

2. Do you think the new options being hidden is enough to let us safely break the interface in the future?

infinisil commented on nixos/modules/security/pam.nix:
````diff
@@ -460,6 +460,94 @@ let
+                    Whether this rule is added to the PAM service config file.
+                  '';
+                };
+                order = mkOption {
+                  type = types.int;
+                  default = 1000;
+                  description = lib.mdDoc ''
+                    Order of this rule in the service file. Rules are arranged in ascending order of this value.
+
+                    ::: {.warning}
+                    The `order` values for the built-in rules are subject to change. If you assign a constant value to this option, a system update could silently reorder your rule. You could be locked out of your system, or your system could be left wide open. If you must use this option, set it to a relative offset from another rule's `order` value:
+
+                    ```nix
+                    {
+                      security.pam.services.login.types.auth.rules.foo.order =
+                        config.security.pam.services.login.types.auth.rules.unix.order + 1;
````
I think we can have an `order`-based implementation for now, but the current one is too unsafe, because it orders rules arbitrarily if they have the same `order` value. To make this safer, an error should be thrown in such cases, therefore requiring definitions to provide one. There also shouldn't be a default order then.

For the future: The problem of such number based orderings is that it requires a sort of centralised registry to decide which modules should use which numbers, which then doesn't work well as soon as you have third-party modules.

This is what a DAG-based ordering solves, and I agree with others that we should really have that in the future. In fact the docs here suggesting the use of other attributes `order` value is almost the same thing. Note that I have [a PR](https://github.com/NixOS/nixpkgs/pull/97392) open to add a generic DAG-based type to Nixpkgs, though I don't have the time to pick it up right now, would appreciate others to do that.

Majiir replied:
Removed the default order and added a check for equal order values in ee5dd1dd40fa77339898580bb614db920b5af46d.

infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
+                  '';
+                };
+                order = mkOption {
+                  type = types.int;
```
Here's an idea to avoid the problem of not reserving enough space between rules :)
```suggestion
                  type = types.float;
```

Majiir replied:
:laughing: That is clever, but I prefer our security infrastructure not rely on floating point comparisons. I will increase the integer spacing. Even if someone needs to add hundreds of rules (unlikely!) they can use `include`/`substack` controls.

Majiir replied:
Increased spacing to 100 in ee5dd1dd40fa77339898580bb614db920b5af46d.

infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -470,90 +558,108 @@ let
+        formatRules = type: pipe cfg.types.${type}.rules [
+          attrValues
+          (filter (rule: rule.enable))
+          (sort (a: b: a.order < b.order))
+          (map (rule: concatStringsSep " " (
+            [ type rule.control rule.modulePath ]
+            ++ map formatModuleArgument rule.args
+          )))
```
It would be really nice to render the `order` of each rule above itself, so that one can just look at `/etc/pam.d/*` to see the values.

Majiir replied:
Added `order` comments in 4995e085152bde3719f01affef1c5a1903d483bf.

infinisil reviewed (changes requested):
Very nice PR! I love the individual commits, makes this very reviewable! The overall idea is solid, though I made some suggestions below.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
+                    Either the full filename of the PAM to be used by the application (it begins with a '/'), or a relative pathname from the default module location. See `module-path` in {manpage}`pam.conf(5)` for details.
+                  '';
+                };
+                args = mkOption {
+                  type = types.listOf types.str;
+                  description = lib.mdDoc ''
+                    Tokens that can be used to modify the specific behavior of the given PAM. Such arguments will be documented for each individual module. See `module-arguments` in {manpage}`pam.conf(5)` for details.
+
+                    Escaping rules for spaces and square brackets are automatically applied.
+
+                    {option}`settings` are automatically added as {option}`args`. It's recommended to use the {option}`settings` option whenever possible so that arguments can be overridden.
+                  '';
+                };
```
`args` is a must-have because that's the contract exposed to PAMs. The string formatting rules apply to all modules. `settings` is an optional abstraction we put on top because it makes modifying the configuration quite nice in Nix.

I had tried combining them by allowing `settings."service = systemd-user" = true` to render without the whitespace escaping. This worked for that example, but it doesn't work when there are derivation paths in the string. Splitting `args` and `settings` turned out to be more natural.

infinisil replied:
The problem with any list-based option is that it can't be overridden really. But if that's what PAM allows then we can't avoid it :+1:

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
+        description = lib.mdDoc ''
+          PAM rules for this service.
+
+          Attribute keys are the `type` of the rules (one of `account`, `auth`, `password`, `session`).
+        '';
+        type = types.attrsOf (types.submodule {
```
My concern was that this would result in repetition in the manual. I guess there wouldn't be _that_ many repeated options, though. We can put it on my wishlist for the module system to support reusable types that are documented in one place in the manual. :smile:

infinisil replied:
Indeed that would be nice, but for now I think it's fine with the duplication in the manual. This does have the benefit of making the documentation more obvious, and it also makes sure you can't have a typo in the key.

Majiir replied:
Fixed in d6bb805932ae8fdea7e516f40b2455109d4b4013.

infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
+                  '';
+                };
+                settings = mkOption {
+                  type = types.attrs;
```
I guess it would be convenient and I see no problem with it. Note that you can use `types.oneOf` then.

Majiir replied:
Fixed in 930937de8343b966553685c576deb83408dd6c37.

Majiir committed:
nixos/pam: add order comment to each rule line

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
 
+      types = mkOption {
+        # This option is experimental and subject to breaking changes without notice.
+        visible = false;
```
Added in d6bb805932ae8fdea7e516f40b2455109d4b4013.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -470,90 +558,108 @@ let
+          (let yubi = config.security.pam.yubico; in { name = "yubico"; enable = cfg.yubicoAuth; control = yubi.control; modulePath = "${pkgs.yubico-pam}/lib/security/pam_yubico.so"; settings = {
+            inherit (yubi) mode debug;
+            chalresp_path = yubi.challengeResponsePath;
+            id = if yubi.mode == "client" then yubi.id else null;
```
Fixed in 930937de8343b966553685c576deb83408dd6c37.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
+          Attribute keys are the `type` of the rules (one of `account`, `auth`, `password`, `session`).
+        '';
+        type = types.attrsOf (types.submodule {
+          options.rules = mkOption {
```
Removed the extra option. The format is now `security.pam.services.<service>.rules.<type>.<rule>`.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -460,6 +460,94 @@ let
         description = lib.mdDoc "Contents of the PAM service file.";
       };
 
+      types = mkOption {
```
Renamed to `rules`.

Majiir commented:
@infinisil Thank you for the review! This is ready for another pass.

infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -470,90 +578,116 @@ let
+      text = let
+        ensureUniqueOrder = type: rules:
+          let
+            headRules = if rules == [] then [] else [ (head rules) ];
+            tailRules = if rules == [] then [] else tail rules;
```
```suggestion
            headRules = take 1 rules;
            tailRules = drop 1 rules;
```

Majiir replied:
Updated in 077cdcc7e9a9ed406860e97ce3590fba15aed18b.

infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -470,90 +578,116 @@ let
+          (map (rule: concatStringsSep " " (
+            [ type rule.control rule.modulePath ]
+            ++ map formatModuleArgument rule.args
+            ++ [ "# ${toString rule.order}" ]
```
```suggestion
            ++ [ "# ${rule.name} (order ${toString rule.order}") ]
```

Majiir replied:
Added in 43f7cb4a95436d58a0fe40ff501d42085d676691.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -18,6 +104,28 @@ let
+        description = lib.mdDoc ''
+          PAM rules for this service.
+
+          ::: {.warning}
+          This option and its suboptions are experimental and subject to breaking changes without notice.
```
What is the purpose if the options are not visible? The descriptions are not rendered in the manual. My thought was to make the warning prominent in the source code, since that's the only place these options can be seen.

infinisil replied:
Ah I see. I guess it's fine then as is.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -6,6 +6,92 @@
+          '';
+        };
+        settings = mkOption {
+          type = with types; attrsOf (nullOr (oneOf [ bool str int path package ]));
```
I'm not intimately familiar with these types, but the use case is to coerce derivations to store paths. For example:

```nix
{
  security.pam.services.sudo.rules.auth = {
    rssh = {
      # ...
      settings.authorized_keys_command = pkgs.writeShellScript "print_user_authorized_keys" ''
        cat "/etc/ssh/authorized_keys.d/$(whoami)"
      '';
    };
  };
}
```

infinisil replied:
I think `types.pathInStore` should be used then. I'd not include `types.path`.

Majiir replied:
Fixed in 5b8439f9666d3fc6ef3139ce48f4d9b34e1fd525. The type is now `attrsOf (nullOr (oneOf [ bool str int pathInStore ]))`.

Majiir commented on nixos/modules/security/pam.nix:
```diff
@@ -6,6 +6,92 @@
+          description = lib.mdDoc ''
+            Order of this rule in the service file. Rules are arranged in ascending order of this value.
+
+            ::: {.warning}
+            The `order` values for the built-in rules are subject to change. If you assign a constant value to this option, a system update could silently reorder your rule. You could be locked out of your system, or your system could be left wide open. When using this option, set it to a relative offset from another rule's `order` value:
```
Not strictly necessary, but I think it's good to point out why fixed order values can be dangerous.

infinisil reviewed (approved):
Looking good!

infinisil commented:
For the record: I'm still not entirely happy about the `order` option, but I don't think it matters very much, these options will be marked as experimental, and it looks like you intend to continue with this effort after this is merged @Majiir.

I think it's okay to merge this as is, but let's give @rissson and @kwohlfahrt some more time to give their approval or suggestions too.

rissson commented:
This is already a massive improvement as you now don't have to override the whole generated file to get what you need.

> For the record: I'm still not entirely happy about the `order` option, but I don't think it matters very much, these options will be marked as experimental

The main issue is being sure that the order is always the same no matter what. As a user, saying that rule A needs to be after rule B is not sufficient enough. Perhaps the `order` being described as an int is not enough. We could also imagine asking the user for an ordered list that contains the order they want their rules in.

I'm very happy to see this making some progress, and am excited to see the next steps!

SuperSandro2000 commented:
I am here because I just hit a merge conflict on rebasing my fork where I changed the pam module rather than using options because the old module just didn't allow it. I just wanted to say a huge thank you to everyone involved that helped closing this giant gap! ❤️

mjm commented:
I've been eagerly awaiting this PR landing in unstable so I could reorder my swaylock auth lines and make both fprintd and password work, thank you for the work that went into this!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/set-the-sudo-password-separately-than-the-login-password-on-nixos/56565/3

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-do-you-want-from-pam-security-pam-in-nixos/33265/19

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/swaylock-u2f-how-to-get-working-fallback/45352/2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/design-preview-pam-rule-ordering-and-targets/66399/1
